pkg/spade-spike/spade-spike.go

```diff
@@ -0,0 +1,286 @@
+}
+
+func (m MessageReceiver) ReceiveError(err error) {
+	m.BSClient.ReceiveError(err)
+}
+
+func (m MessageReceiver) PeerConnected(id peer.ID) {
+	m.BSClient.PeerConnected(id)
+}
+func (m MessageReceiver) PeerDisconnected(id peer.ID) {
+	m.BSClient.PeerDisconnected(id)
+}
+
+type BitswapClient struct {
+	host    host.Host
+	timeout time.Duration
+}
+
+func NewBitswapClient(host host.Host, timeout time.Duration) BitswapClient {
+	return BitswapClient{
+		host:    host,
+		timeout: timeout,
+	}
+}
+
+// Attempts to decode the block data into a node and return its links
+func FindLinks(ctx context.Context, blk blocks.Block) ([]datamodel.Link, error) {
+	if blk.Cid().Prefix().Codec == goCid.Raw {
+		// This can happen at the bottom of the tree
+		return nil, errors.New("raw block encountered " + blk.Cid().String())
+	}
+
+	decoder, err := cidlink.DefaultLinkSystem().DecoderChooser(cidlink.Link{Cid: blk.Cid()})
+
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := ipld.Decode(blk.RawData(), decoder)
+	if err != nil {
+		return nil, err
+	}
+
+	links, err := traversal.SelectLinks(node)
+	if err != nil {
+		return nil, err
+	}
+
+	return links, nil
+}
+
+// Returns the raw block data, the links, and error if any
+func (c BitswapClient) RetrieveBlock(
+	parent context.Context,
+	target peer.AddrInfo,
+	cid goCid.Cid) (blocks.Block, error) {
+	network := bsnet.NewFromIpfsHost(c.host, SingleContentRouter{
+		AddrInfo: target,
+	})
+	bswap := bsclient.New(parent, network, blockstore.NewBlockstore(datastore.NewMapDatastore()))
+	notFound := make(chan struct{})
+	network.Start(MessageReceiver{BSClient: bswap, MessageHandler: func(
+		ctx context.Context, sender peer.ID, incoming bsmsg.BitSwapMessage) {
+		if sender == target.ID && slices.Contains(incoming.DontHaves(), cid) {
+			logger.Info("Block not found")
+			close(notFound)
+		}
+	}})
+	defer bswap.Close()
+	defer network.Stop()
+	connectContext, cancel := context.WithTimeout(parent, c.timeout)
+	defer cancel()
+	logger.Info("Connecting to target peer...")
+	err := c.host.Connect(connectContext, target)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to target peer, %s", err)
+	}
+
+	resultChan := make(chan blocks.Block)
+	errChan := make(chan error)
+	go func() {
+		logger.Debug("Retrieving block...")
+		blk, err := bswap.GetBlock(connectContext, cid)
+		if err != nil {
+			logger.Info(err)
+			errChan <- err
+		} else {
+			resultChan <- blk
+		}
+	}()
+	select {
+	case <-notFound:
+		return nil, errors.New("DONT_HAVE received from the target peer")
+
+	case blk := <-resultChan:
+		return blk, nil
+
+	case err := <-errChan:
+		return nil, fmt.Errorf("error received %s", err)
+	}
+}
+
+// Starts with the root CID, then fetches a random CID from the children and grandchildren nodes
+// Returns true if all retrievals were successful, false if any failed
+func SpadeTraversal(ctx context.Context, startingCid goCid.Cid, p peer.AddrInfo) (bool, error) {
```

Comment:
@ribasushi does this logic make sense to you?

Reply:
```
// It will always be 3 layers, no more no less
// ref: https://github.com/filecoin-project/go-dagaggregator-unixfs#grouping-unixfs-structure
```

this is inaccurate: this format is utterly obsolete and we are only using it because... inertia. So hardcoding this within the function called `SpadeTraversal` is... not cool.

Instead make the maxDeptha parameter to the function, and call it with 4 ( not 3, you want to fetch the actual thing itself too ) for the case of daghaus, and only for the initial subset we are doing (w3up daghaus contracts will not be verifiable in this manner at all )

Reply:
Good callout - fixed! Added a `traverseDepth` param here. I refer to the first layer of the tree (root node) as `depth: 0` , so passing in `traverseDepth: 0` is still a valid operation, only fetching once. In this case we still need to pass `3` since those data CIDs are at depth 3. I did this way because starting at `1` would make `traverseDepth: 0` a noop, but let me know if this is a strange convention, easy to change!